File: appmonet-sdk/appmonet-sdk-mopub/src/main/java/com/monet/bidder/MonetInterstitialListener.java
package com.monet.bidder;

import android.content.Context;
import android.content.Intent;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.monet.bidder.threading.InternalRunnable;
import com.mopub.common.logging.MoPubLog;
import com.mopub.mobileads.AdLifecycleListener;
import com.mopub.mobileads.MoPubErrorCode;

import java.lang.ref.WeakReference;

import static com.monet.bidder.Constants.APPMONET_BROADCAST;
import static com.monet.bidder.Constants.APPMONET_BROADCAST_MESSAGE;
import static com.monet.bidder.CustomEventInterstitial.ADAPTER_NAME;
import static com.mopub.common.logging.MoPubLog.AdLogEvent.CLICKED;
import static com.mopub.common.logging.MoPubLog.AdLogEvent.LOAD_FAILED;
import static com.mopub.common.logging.MoPubLog.AdLogEvent.LOAD_SUCCESS;


/**
 * Created by jose on 3/14/18.
 */

class MonetInterstitialListener implements AdServerBannerListener {
  private static final Logger sLogger = new Logger("MonetInterstitialListener");

  private final WeakReference<Context> context;
  private final SdkManager sdkManager;
  private final AdLifecycleListener.LoadListener mLoadListener;
  private final AdLifecycleListener.InteractionListener mInteractionListener;
  private final String adUnitId;


  public MonetInterstitialListener(@Nullable AdLifecycleListener.LoadListener mLoadListener,
                                   @Nullable AdLifecycleListener.InteractionListener mInteractionListener,
                                   @NonNull String adUnitId,
                                   @NonNull Context context, @NonNull SdkManager sdkManager) {
    this.sdkManager = sdkManager;
    this.mLoadListener = mLoadListener;
    this.mInteractionListener = mInteractionListener;
    this.adUnitId = adUnitId;
    this.context = new WeakReference<>(context);
  }

  @Override
  public void onAdClosed() {
    if (this.context.get() != null) {
      LocalBroadcastManager.getInstance(this.context.get()).sendBroadcast(
          new Intent(APPMONET_BROADCAST)
              .putExtra(APPMONET_BROADCAST_MESSAGE,
                  "interstitial_dismissed"));
    }
  }

  @Override
  public void onAdOpened() {
    //no implementation
  }

  @Override
  public boolean onAdLoaded(View view) {
    try {
      sdkManager.getUiThread().run(new InternalRunnable() {
        @Override
        public void runInternal() {
          MoPubLog.log(adUnitId, LOAD_SUCCESS, ADAPTER_NAME);
          if (mLoadListener != null) {
            mLoadListener.onAdLoaded();
          }
        }

        @Override
        public void catchException(Exception e) {
          sLogger.warn("failed to finish on view: ", e.getMessage());
          onAdError(ErrorCode.INTERNAL_ERROR);
        }
      });
    } catch (Exception e) {
      sLogger.error("error while loading into MoPub", e.getMessage());
      onAdError(ErrorCode.INTERNAL_ERROR);
      return false;
    }
    return true;
  }

  @Override
  public void onAdClicked() {
    MoPubLog.log(adUnitId, CLICKED, ADAPTER_NAME);
    if (mInteractionListener != null) {
      mInteractionListener.onAdClicked();
    }
  }

  @Override
  public void onAdError(ErrorCode errorCode) {
    MoPubLog.log(LOAD_FAILED, ADAPTER_NAME, moPubErrorCode(errorCode).getIntCode(),
        moPubErrorCode(errorCode));
    if (mInteractionListener != null) {
      mInteractionListener.onAdFailed(moPubErrorCode(errorCode));
    }
  }

  @Override
  public void onAdRefreshed(View view) {
    //no implementation
  }

  private static MoPubErrorCode moPubErrorCode(ErrorCode errorCode) {
    switch (errorCode) {
      case INTERNAL_ERROR:
        return MoPubErrorCode.INTERNAL_ERROR;
      case NO_FILL:
        return MoPubErrorCode.NETWORK_NO_FILL;
      case TIMEOUT:
        return MoPubErrorCode.NETWORK_TIMEOUT;
      case BAD_REQUEST:
        return MoPubErrorCode.NETWORK_INVALID_STATE;
      default:
        return MoPubErrorCode.UNSPECIFIED;
    }
  }
}

File: appmonet-sdk/appmonet-sdk-mopub/src/main/java/com/monet/bidder/MonetInterstitialListener.kt
package com.monet.bidder

import android.content.Context
import android.content.Intent
import android.view.View
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.monet.AdServerBannerListener
import com.monet.AdServerBannerListener.ErrorCode
import com.monet.AdServerBannerListener.ErrorCode.BAD_REQUEST
import com.monet.AdServerBannerListener.ErrorCode.INTERNAL_ERROR
import com.monet.AdServerBannerListener.ErrorCode.NO_FILL
import com.monet.AdServerBannerListener.ErrorCode.TIMEOUT
import com.monet.bidder.Constants.APPMONET_BROADCAST
import com.monet.bidder.Constants.APPMONET_BROADCAST_MESSAGE
import com.mopub.common.logging.MoPubLog
import com.mopub.common.logging.MoPubLog.AdLogEvent.CLICKED
import com.mopub.common.logging.MoPubLog.AdLogEvent.LOAD_FAILED
import com.mopub.common.logging.MoPubLog.AdLogEvent.LOAD_SUCCESS
import com.mopub.mobileads.AdLifecycleListener.InteractionListener
import com.mopub.mobileads.AdLifecycleListener.LoadListener
import com.mopub.mobileads.MoPubErrorCode
import com.mopub.mobileads.MoPubErrorCode.NETWORK_INVALID_STATE
import com.mopub.mobileads.MoPubErrorCode.NETWORK_NO_FILL
import com.mopub.mobileads.MoPubErrorCode.NETWORK_TIMEOUT
import com.mopub.mobileads.MoPubErrorCode.UNSPECIFIED
import java.lang.ref.WeakReference

/**
 * Created by jose on 3/14/18.
 */
internal class MonetInterstitialListener(
  private val mLoadListener: LoadListener?,
  private val mInteractionListener: InteractionListener?,
  context: Context,
  private val sdkManager: SdkManager
) : AdServerBannerListener<View?> {
  private val context: WeakReference<Context?> = WeakReference(context)
  var adUnitId: String? = null
  override fun onAdClosed() {
    if (context.get() != null) {
      LocalBroadcastManager.getInstance(context.get()!!).sendBroadcast(
          Intent(APPMONET_BROADCAST)
              .putExtra(
                  APPMONET_BROADCAST_MESSAGE,
                  "interstitial_dismissed"
              )
      )
    }
  }

  override fun onAdOpened() {
    //no implementation
  }

  override fun onAdLoaded(view: View?): Boolean {
    try {
      sdkManager.uiThread.run {
        try {
          MoPubLog.log(adUnitId, LOAD_SUCCESS, CustomEventInterstitial.ADAPTER_NAME)
          mLoadListener?.onAdLoaded()
        } catch (e: Exception) {
          sLogger.warn("failed to finish on view: ", e.message)
          onAdError(INTERNAL_ERROR)
        }
      }
    } catch (e: Exception) {
      sLogger.error("error while loading into MoPub", e.message)
      onAdError(INTERNAL_ERROR)
      return false
    }
    return true
  }

  override fun onAdClicked() {
    MoPubLog.log(adUnitId, CLICKED, CustomEventInterstitial.ADAPTER_NAME)
    mInteractionListener?.onAdClicked()
  }

  override fun onAdError(errorCode: ErrorCode) {
    MoPubLog.log(
        LOAD_FAILED, CustomEventInterstitial.ADAPTER_NAME, moPubErrorCode(errorCode).intCode,
        moPubErrorCode(errorCode)
    )
    mInteractionListener?.onAdFailed(moPubErrorCode(errorCode))
  }

  override fun onAdRefreshed(view: View?) {
    //no implementation
  }

  companion object {
    private val sLogger = Logger("MonetInterstitialListener")
    private fun moPubErrorCode(errorCode: ErrorCode): MoPubErrorCode {
      return when (errorCode) {
        INTERNAL_ERROR -> MoPubErrorCode.INTERNAL_ERROR
        NO_FILL -> NETWORK_NO_FILL
        TIMEOUT -> NETWORK_TIMEOUT
        BAD_REQUEST -> NETWORK_INVALID_STATE
        else -> UNSPECIFIED
      }
    }
  }

  override fun onLeftApplication() {
    TODO("Not yet implemented")
  }

}